sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+
+This is not the same as microservices, is just to separate the view from the API logic.
+
+This is neither the same as Progressive Web Apps (PWA), since in these kind of applications, the content is being loaded and served from the browser self, so even without an internet connection, the application works and can serve the content has been already loaded and sync when a internet connection is available again. By the other side, in the API approach, a network connection is mandatory for the correct operation of the API and its components.
+
+In the API approach, one of the things to have present is that APIs are decoupled from the view. So we can use and switch them between different platforms. Some usual components to achieve this approach are:
+
+- API using some backend programming language such as Java for example.
+- SPA (Single Page Application) is the frontend in which we can use JS, CSS, HTML5 or some JS framework such as React, Angular or Vue since the're some of the most popular SPA options right now.
+
+
+#### Dependency injection concepts
+
+A simplification and better practice to use dependency injection in Spring was talked about. We were changing from field injection towards constructor injection.
+
+##### Field injection
+
+Field injection allows to link resources, so for example in the next code snippet, a controller is asking to Spring to `@Autowire` the `BookService` class from its pool of registered services:
+
+```java
+import org.springframework.beans.factory.annotation.Autowired;
+
+@RestController
+@RequestMapping("/api/book")
+public class BookController {
+
+    @Autowired
+    private BookService bookService;
+    // ...
+}
+```
+
+__NOTE__: Field injection is the best choise for JUnit tests.
+
+
+##### Constructor injection
+
+Other way to use our service into our controller is to create a constructor that receives the service and populate it into a local field.
+
+So, once we create the constructor, we can declare `BookService` field as final (by using `@Autowired` that isn't possible). And even, we don't need anymore to use `@Autowired` annotation at all as is showed following:
+
+```java
+@RestController
+@RequestMapping("/api/book")
+public class BookController {
+
+    private final BookService bookService;
+
+    BookController(BookService bookService) {
+        this.bookService = bookService;
+    }
+
+    // ...
+}
+```
+
+__NOTE__: Constructor injection is the best choice to use in main code.
+
+__NOTE__: When we declare a field as final, it cannot be modified after, thus we have quicker applications and less prone to modifications by mistake in the fields we don't want. By other side, is not recommended use final keyword in classes and methods because that lead to applications that are hard to work with.
```

Comment:
I would clarify here that it's not recommended to use the `final` keyword *in front of class or method declarations*.
If you have final fields, it is strictly speaking used "in classes".